src/inmanta/deploy/scheduler.py

```diff
@@ -606,6 +626,21 @@ async def is_agent_running(self, name: str) -> bool:
         """
         return name in self._workers and self._workers[name].is_running()
 
+    async def repair_resource(self, resource: ResourceIdStr, repair_timer: int) -> None:
+        async with self._scheduler_lock:
```

Comment:
it is somewhat interesting that the only async thing here is the lock.

I wonder, can we safely do a deploy, outside the lock, because we know we are sitting on the IOloop thread? 

It would simplify the timers somewhat I think

Reply:
> I wonder, can we safely do a deploy, outside the lock, because we know we are sitting on the IOloop thread?

I don't know. Does this remark also apply to other places where we only do sync operations under the async lock ? e.g. regular client triggered [deploy](https://github.com/inmanta/inmanta-core/blob/dbf5c7ae7233a775375b50e77dc76cdb2e1bd68d/src/inmanta/deploy/scheduler.py#L310) ?

@sanderr do you know ?

Reply:
it is very very specific to this code path

Reply:
> Does this remark also apply to other places where we only do sync operations under the async lock ?

Only if **all** places would only have sync operations. i.e. if one place exists that does an `await` under the lock, then it might pass control to another async function. If that one at some point does a bunch of sync stuff that should be under the lock, it affects the integrity of the original location.

And I honestly think that that same argument applies here, doesn't it @wouterdb?

Reply:
Just realized that requires-provides is mutable, so this is complicated as hell, comment below probably doesn't apply, you are right, lock it required 

---- old comment -> ignore --- 

Good point, so the question becomes: is any of the code below the lock affected by this.

Interestingly the  ScheduledWork is entirely sync, so it can be considered to be atomic wrt to asyncio. We never read from it (except de-queue). So data corruption is not a risk, nor is read-confusion, only re-ordering of operations.  All Operations are idempotent, associative and commutative **except** for de-queue, but that only happens in the agent worker (so the IOLoop).

We may be able to do without the lock, but that is perhaps not for now.

Reply:
> ---- old comment -> ignore ---

The part above or below this marker? EDIT: nevermind, the comment above references the other.